packages/grafana-ui/src/components/Combobox/useLatestAsyncCall.ts

```diff
@@ -0,0 +1,43 @@
+import { useCallback, useRef } from 'react';
+
+type AsyncFn<T, V> = (value: T) => Promise<V>;
+
+/**
+ * Wraps an async function to ensure that only the latest call is resolved.
+ * Used to prevent a faster call being overwritten by an earlier slower call.
+ */
+export function useLatestAsyncCall<T, V>(fn: null | AsyncFn<T, V>): AsyncFn<T, V> {
+  const latestValue = useRef<T>();
+
+  const wrappedFn = useCallback(
+    (value: T) => {
+      if (!fn) {
+        throw new Error('useLatestAsyncCall was called with a null function');
```

Comment:
I find this a bit confusing and took me a while to figure out why we allow `null` values for the `fn` which is because you cannot have conditional hooks right? In that case wouldn't it make sense to just make non async values async? Ie wrap options in `Promise.resolve([])`? I would think the perf hit would be negligible and it could clean up the code as there would be no need for any `isAsync` check in the code later on.

Reply:
yeah, you're right. that's probably a less worse way to check this. 

The smell should have been obvious to me that I was changing the implementation of this hook to work around a detail of the consumer